File: read_gdoc.py
"""
Read Google Doc (.gdoc) files from Google Drive for Desktop.

Usage:
  python read_gdoc.py "G:/My Drive/path/to/file.gdoc"
  python read_gdoc.py --doc-id 1BDuu4uj4K5VKpEuitHz4wYeNxx230aGfKgUewAGbIxg

First-time setup:
  1. Go to https://console.cloud.google.com/
  2. Create a project (or select existing)
  3. Enable the Google Docs API: APIs & Services > Enable APIs > search "Google Docs API" > Enable
  4. Create OAuth credentials: APIs & Services > Credentials > Create Credentials > OAuth client ID
     - Application type: Desktop app
     - Download the JSON file
  5. Save as C:/Projects/ExpenseTracker/gdoc_credentials.json
  6. Run this script - it will open a browser for one-time authorization
"""

import sys
import os
import json
import sqlite3
import re

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
CREDENTIALS_FILE = os.path.join(SCRIPT_DIR, "gdoc_credentials.json")
TOKEN_FILE = os.path.join(SCRIPT_DIR, "gdoc_token.json")
DRIVEFSDIR = os.path.join(os.environ["LOCALAPPDATA"], "Google", "DriveFS")
SCOPES = ["https://www.googleapis.com/auth/documents.readonly"]


def find_drivefsdb():
    """Find the DriveFS metadata database."""
    if not os.path.isdir(DRIVEFSDIR):
        return None
    for entry in os.listdir(DRIVEFSDIR):
        db_path = os.path.join(DRIVEFSDIR, entry, "metadata_sqlite_db")
        if os.path.isfile(db_path):
            return db_path
    return None


def get_doc_id_from_gdoc(gdoc_path):
    """Extract the Google Doc ID from the DriveFS metadata database."""
    filename = os.path.basename(gdoc_path)
    db_path = find_drivefsdb()
    if not db_path:
        print("ERROR: Could not find DriveFS metadata database.", file=sys.stderr)
        sys.exit(1)

    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()
    cursor.execute(
        "SELECT id FROM items WHERE local_title = ? AND mime_type = 'application/vnd.google-apps.document'",
        (filename,)
    )
    row = cursor.fetchone()
    conn.close()

    if not row:
        # Try without extension
        base = os.path.splitext(filename)[0]
        conn = sqlite3.connect(db_path)
        cursor = conn.cursor()
        cursor.execute(
            "SELECT id FROM items WHERE local_title LIKE ? AND mime_type = 'application/vnd.google-apps.document'",
            (f"%{base}%",)
        )
        row = cursor.fetchone()
        conn.close()

    if not row:
        print(f"ERROR: Could not find doc ID for '{filename}' in DriveFS metadata.", file=sys.stderr)
        sys.exit(1)

    return row[0]


def get_credentials():
    """Get or refresh OAuth credentials."""
    from google.oauth2.credentials import Credentials
    from google_auth_oauthlib.flow import InstalledAppFlow
    from google.auth.transport.requests import Request

    creds = None

    if os.path.exists(TOKEN_FILE):
        creds = Credentials.from_authorized_user_file(TOKEN_FILE, SCOPES)

    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            if not os.path.exists(CREDENTIALS_FILE):
                print("ERROR: gdoc_credentials.json not found.", file=sys.stderr)
                print("", file=sys.stderr)
                print("Setup instructions:", file=sys.stderr)
                print("  1. Go to https://console.cloud.google.com/", file=sys.stderr)
                print("  2. Create/select a project", file=sys.stderr)
                print("  3. Enable the Google Docs API", file=sys.stderr)
                print("  4. Create OAuth credentials (Desktop app)", file=sys.stderr)
                print(f"  5. Save the JSON as: {CREDENTIALS_FILE}", file=sys.stderr)
                sys.exit(1)

            flow = InstalledAppFlow.from_client_secrets_file(CREDENTIALS_FILE, SCOPES)
            creds = flow.run_local_server(port=0)

        with open(TOKEN_FILE, "w") as f:
            f.write(creds.to_json())

    return creds


def read_google_doc(doc_id):
    """Fetch the content of a Google Doc as plain text."""
    from googleapiclient.discovery import build

    creds = get_credentials()
    service = build("docs", "v1", credentials=creds)
    doc = service.documents().get(documentId=doc_id).execute()

    text_parts = []
    for element in doc.get("body", {}).get("content", []):
        if "paragraph" in element:
            for pe in element["paragraph"].get("elements", []):
                if "textRun" in pe:
                    text_parts.append(pe["textRun"]["content"])

    return "".join(text_parts)


def main():
    if len(sys.argv) < 2:
        print("Usage: python read_gdoc.py <path-to.gdoc>", file=sys.stderr)
        print("       python read_gdoc.py --doc-id <DOCUMENT_ID>", file=sys.stderr)
        sys.exit(1)

    if sys.argv[1] == "--doc-id":
        doc_id = sys.argv[2]
    else:
        gdoc_path = sys.argv[1]
        doc_id = get_doc_id_from_gdoc(gdoc_path)

    text = read_google_doc(doc_id)
    print(text)


if __name__ == "__main__":
    main()

File: read_gdoc.js
const fs = require("fs");
const path = require("path");
const http = require("http");
const { google } = require("googleapis");

const CREDS_PATH = path.join(__dirname, "documentation", "client_secret.json");
const TOKEN_PATH = path.join(__dirname, "documentation", "token.json");
const SCOPES = [
  "https://www.googleapis.com/auth/documents.readonly",
  "https://www.googleapis.com/auth/drive.readonly",
];

async function authorize() {
  const creds = JSON.parse(fs.readFileSync(CREDS_PATH, "utf8"));
  const { client_id, client_secret, redirect_uris } = creds.installed;
  const oAuth2Client = new google.auth.OAuth2(client_id, client_secret, "http://localhost:3333");

  if (fs.existsSync(TOKEN_PATH)) {
    const token = JSON.parse(fs.readFileSync(TOKEN_PATH, "utf8"));
    oAuth2Client.setCredentials(token);

    // Check if token needs refresh
    if (token.expiry_date && token.expiry_date < Date.now()) {
      try {
        const { credentials } = await oAuth2Client.refreshAccessToken();
        oAuth2Client.setCredentials(credentials);
        fs.writeFileSync(TOKEN_PATH, JSON.stringify(credentials, null, 2));
      } catch (e) {
        console.error("Token refresh failed, re-authorizing...");
        return getNewToken(oAuth2Client);
      }
    }
    return oAuth2Client;
  }

  return getNewToken(oAuth2Client);
}

function getNewToken(oAuth2Client) {
  return new Promise((resolve, reject) => {
    const authUrl = oAuth2Client.generateAuthUrl({
      access_type: "offline",
      scope: SCOPES,
    });

    console.log("\n=== Google Authorization Required ===");
    console.log("Open this URL in your browser:\n");
    console.log(authUrl);
    console.log("\nWaiting for authorization...\n");

    const server = http.createServer(async (req, res) => {
      try {
        const url = new URL(req.url, "http://localhost:3333");
        const code = url.searchParams.get("code");
        if (!code) {
          res.writeHead(400, { "Content-Type": "text/html" });
          res.end("<h1>No code received</h1>");
          return;
        }

        const { tokens } = await oAuth2Client.getToken(code);
        oAuth2Client.setCredentials(tokens);
        fs.writeFileSync(TOKEN_PATH, JSON.stringify(tokens, null, 2));

        res.writeHead(200, { "Content-Type": "text/html" });
        res.end("<h1>Authorization successful!</h1><p>You can close this tab.</p>");
        server.close();
        resolve(oAuth2Client);
      } catch (e) {
        res.writeHead(500, { "Content-Type": "text/html" });
        res.end("<h1>Error</h1><p>" + e.message + "</p>");
        server.close();
        reject(e);
      }
    });

    server.listen(3333, () => {
      // Try to open browser automatically
      const { exec } = require("child_process");
      exec(`start "" "${authUrl}"`);
    });
  });
}

function extractDocId(gdocPath) {
  // .gdoc files are JSON with a "doc_id" or "url" field
  const content = fs.readFileSync(gdocPath, "utf8");
  try {
    const json = JSON.parse(content);
    if (json.doc_id) return json.doc_id;
    if (json.url) {
      const match = json.url.match(/\/document\/d\/([a-zA-Z0-9_-]+)/);
      if (match) return match[1];
    }
  } catch (e) {
    // Maybe it's just a URL
    const match = content.match(/\/document\/d\/([a-zA-Z0-9_-]+)/);
    if (match) return match[1];
  }
  throw new Error("Could not extract document ID from: " + gdocPath);
}

async function readGoogleDoc(auth, docId) {
  const docs = google.docs({ version: "v1", auth });
  const res = await docs.documents.get({ documentId: docId });
  const doc = res.data;

  let text = "";
  if (doc.body && doc.body.content) {
    for (const element of doc.body.content) {
      if (element.paragraph) {
        for (const part of element.paragraph.elements) {
          if (part.textRun && part.textRun.content) {
            text += part.textRun.content;
          }
        }
      } else if (element.table) {
        for (const row of element.table.tableRows) {
          const cells = [];
          for (const cell of row.tableCells) {
            let cellText = "";
            for (const cellContent of cell.content) {
              if (cellContent.paragraph) {
                for (const part of cellContent.paragraph.elements) {
                  if (part.textRun && part.textRun.content) {
                    cellText += part.textRun.content.trim();
                  }
                }
              }
            }
            cells.push(cellText);
          }
          text += cells.join(" | ") + "\n";
        }
      }
    }
  }
  return text;
}

async function searchByName(auth, fileName) {
  const drive = google.drive({ version: "v3", auth });
  const safeName = fileName.replace(/'/g, "\\'");
  const res = await drive.files.list({
    q: `name = '${safeName}' and mimeType = 'application/vnd.google-apps.document'`,
    fields: "files(id, name)",
    pageSize: 5,
  });
  if (!res.data.files || res.data.files.length === 0) {
    throw new Error("No Google Doc found with name: " + fileName);
  }
  return res.data.files[0].id;
}

async function main() {
  const input = process.argv[2];
  if (!input) {
    console.error("Usage: node read_gdoc.js <doc-name-or-path>");
    console.error("       node read_gdoc.js --auth   (just authorize)");
    console.error("       node read_gdoc.js --search <name>  (search by name)");
    process.exit(1);
  }

  const auth = await authorize();

  if (input === "--auth") {
    console.log("Authorization successful! Token saved to:", TOKEN_PATH);
    process.exit(0);
  }

  let docId;

  if (input === "--search" || input.endsWith(".gdoc")) {
    // Search by document name via Drive API
    const searchName = input === "--search"
      ? process.argv[3]
      : path.basename(input, ".gdoc");
    console.log("Searching for Google Doc:", searchName);
    docId = await searchByName(auth, searchName);
    console.log("Found doc ID:", docId);
  } else if (input.match(/^[a-zA-Z0-9_-]{20,}$/)) {
    // Direct doc ID
    docId = input;
  } else {
    // Treat as document name
    console.log("Searching for Google Doc:", input);
    docId = await searchByName(auth, input);
    console.log("Found doc ID:", docId);
  }

  const text = await readGoogleDoc(auth, docId);
  console.log("\n--- Document Content ---\n");
  console.log(text);
}

main().catch((err) => {
  console.error("Error:", err.message);
  process.exit(1);
});
